### PR TITLE
Review

### PR DESCRIPTION
**Pull request by wbazant**
- supports .bed12 files that list exons
- script for indexing .fai files with SQLite
- accessing reference sequences uses SQLite files to retrieve stuff from the index
- remove forceStrandedness from the API
- refactoring: some tenured functionality is going out of util and into their own packages
- fix the little scripts for integration testing and add a new one
- remove memswap limit from docker-compose.yml

### CODE REVIEW

File: schema/url/sequence.raml [hunk 1]
```diff
@@ -26,10 +26,6 @@ types:
     type: string
     enum: ['bed', 'gff3']
 
-  SequenceType:
```
Comment (ryanrdoherty) [inline schema/url/sequence.raml:29]: Curious why you changed this to a raw string- seems like this should be a well-defined vocabulary, expanded as we add additional sequence types/sources.

Comment (ryanrdoherty) [reply schema/url/sequence.raml:29]: Looking further, it seems like you've converted the entire setup to be configurable.  Is this really extendable to other types?  I'm asking if all the service config options and the request options are applicable regardless of the sequence type?

File: scripts/index_to_sqlite3.sh [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
```
Comment (ryanrdoherty) [inline scripts/index_to_sqlite3.sh:1]: Looks like the new sqlite files are inputs to this service.  That means they will be generated by the workflow right?  Perhaps that is a better location for this script.

File: src/main/java/org/veupathdb/service/sr/reference/ReferenceDAO.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+      } catch (SQLException | IOException e){
+        throw new RuntimeException(e);
+      }
+    };
+  }
+  private FakeFastaSequenceIndex transientIndex(PreparedStatement statement){
+     Function<String, FastaSequenceIndexEntry> f = contigName -> {
+      try {
+        statement.setString(1, contigName);
+        ResultSet rs = statement.executeQuery();
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/service/sr/reference/ReferenceDAO.java]: Better to put the ResultSet declaration in a try-with-resources

File: src/main/java/org/veupathdb/service/sr/reference/FakeFastaSequenceIndex.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import java.util.Iterator;
+import java.util.Properties;
+import java.util.function.Function;
+import htsjdk.samtools.reference.FastaSequenceIndex;
+import htsjdk.samtools.reference.FastaSequenceIndexEntry;
+
+/*
+ * An adapter converting a contig -> entry function to an index object
+ */
+public class FakeFastaSequenceIndex extends FastaSequenceIndex {
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/service/sr/reference/FakeFastaSequenceIndex.java]: I think this needs a better description about what role this plays in the samtools API and why you need to rig this up.  Also "Fake" is probably not quite the right name modifier; maybe "Stub" or "Facade" as a suffix to imply we are subbing this in instead of the samtools version for API compliance purposes.

File: src/main/java/org/veupathdb/service/sr/response/StreamSequences.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+import htsjdk.samtools.util.Interval;
+import java.nio.charset.Charset;
+
+public class StreamSequences {
+
+  private static final Charset CHARSET = Charset.forName("UTF-8");
+
+  private static final byte[] LINE_SEPARATOR = String.valueOf("\n").getBytes(CHARSET);
+
+  public static void write(OutputStream outputStream, IndexedFastaSequenceFile sequences, List<BEDFeature> features, DeflineFormat deflineFormat, int requestedBasesPerLine){
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/service/sr/response/StreamSequences.java:27]: Curious if we might anticipate the BED feature list to be excessively long?  I know we are planning to use a WDK reporter to deliver a resulting file, so could there be e.g. 1M+ features in the input?  If so, we probably should be saving it off somewhere and streaming them into this code to avoid runtime memory issues.

File: src/main/java/org/veupathdb/service/sr/response/StreamSequences.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    int basesPerLine = requestedBasesPerLine > 0 ? requestedBasesPerLine : Integer.MAX_VALUE;
+    try (var buf = new BufferedOutputStream(outputStream)) {
+      for (var feature : features) {
+        var defline = Deflines.deflineForFeature(feature, deflineFormat);
+        var bases = Bases.getBasesForBedFeature(sequences, feature);
+
+        appendSequenceToStream(buf, defline, bases, basesPerLine);
+        buf.flush();
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/service/sr/response/StreamSequences.java]: If you're using a buffered stream already, no need to flush until the end (after the loop); let the buffer do its job.

File: src/main/java/org/veupathdb/service/sr/response/StreamSequences.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+  private static final Charset CHARSET = Charset.forName("UTF-8");
+
+  private static final byte[] LINE_SEPARATOR = String.valueOf("\n").getBytes(CHARSET);
+
+  public static void write(OutputStream outputStream, IndexedFastaSequenceFile sequences, List<BEDFeature> features, DeflineFormat deflineFormat, int requestedBasesPerLine){
+    int basesPerLine = requestedBasesPerLine > 0 ? requestedBasesPerLine : Integer.MAX_VALUE;
+    try (var buf = new BufferedOutputStream(outputStream)) {
+      for (var feature : features) {
+        var defline = Deflines.deflineForFeature(feature, deflineFormat);
+        var bases = Bases.getBasesForBedFeature(sequences, feature);
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/service/sr/response/StreamSequences.java:32]: Again, wondering how big this byte[] is expected to be?

File: src/main/java/org/veupathdb/service/sr/response/StreamSequences.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+  /*
+   * appends sequence to stream
+   * customized from htsjdk.samtools.reference.FastaReferenceWriter;
+   */
+  private static void appendSequenceToStream(OutputStream outputStream, String defline, byte[] bases, int basesPerLine) throws IOException {
+    outputStream.write(defline.getBytes(CHARSET));
+    outputStream.write(LINE_SEPARATOR);
+
+    final int to = bases.length;
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/service/sr/response/StreamSequences.java]: This code is a little hard to follow (even though its intention is clear).  I think some better-named variables would clarify things.  Like what is next vs nextLength?  "to" is never modified, so maybe change to just "totalBaseCount", etc.

File: src/main/java/org/veupathdb/service/sr/service/SequenceRetrievalAsyncService.java [hunk 9]
```diff
@@ -54,60 +48,71 @@ private JobResponse asyncResponse(
...
+      case URL:
+        spec.setFeaturesUrl(url);
+        break;
+    }
+    return PostSequencesAsyncBySequenceTypeAndFileFormatResponse.respond200WithApplicationJson(asyncResponse(spec));
+  }
 
-    } catch (IOException e) {
-      throw new RuntimeException("Unable to complete file processing", e);
+  private static String slurp(File file){
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/service/sr/service/SequenceRetrievalAsyncService.java:95]: Worried about file size again here.  If it is expected to be small, maybe we can do a quick size check before reading into a string?

File: src/main/java/org/veupathdb/service/sr/service/SequenceRetrievalAsyncService.java [hunk 10]
```diff
@@ -54,60 +48,71 @@ private JobResponse asyncResponse(
...
+    try {
+      return getFileContent(new FileInputStream(file), "UTF-8");
+    } catch (IOException e){
+      throw new RuntimeException(e);
     }
   }
+
+  //https://stackoverflow.com/a/15161590
```
Comment (ryanrdoherty) [inline src/main/java/org/veupathdb/service/sr/service/SequenceRetrievalAsyncService.java:103]: You don't need to write this code.  If you're not concerned about size and want to read everything into a String, this is simpler:
```
ByteArrayOutputStream out = new ByteArrayOutputStream();
in.transferTo(out);
return out.toString();
```

File: scripts/index_to_sqlite3.sh (hunk 2 above)
Comment (ryanrdoherty) [reply scripts/index_to_sqlite3.sh:1]: Ok, I see this is a convenience to generate sqlite index from the existing test fai.  Wonder if you could add comments about why this is so simple?  I can guess that the fai is tab-delimited and thus can be read directly into the table as long as the column order lines up, but can you add comment to confirm this is what's happening?

**Review by ryanrdoherty — APPROVED** [review]
Overall, looks pretty good.  Curious about sqlite performance, which we'll find out shortly.  Comments of most concern are around loading up too much request-scoped data into memory at one time, esp if bed features are "long" (100 chars or more?) times a WDK transcript result size (could be 1M+).  Even if they're short, it would be nice to be streaming this stuff since we can (no processing requires everything in memory at once).   These memory concerns should be addressed before pushing to production but no reason this should not go out to dev as-is.  Documentation comments could be addressed later as well, though often nice to add while everything is fresh in your mind.

File: schema/url/sequence.raml (hunk 1 above)
Comment (wbazant) [reply schema/url/sequence.raml:29]: I do think the new sequences should be added with a config change, not code change - it's how we're setting up the service, e.g one genomic and one protein sequence in test, and a few different ones in production.

The options are "is stranded" which tells the service what to do about negative strand entries in .bed. The other options limits for throughput we could potentially have them global rather than different for each request!

File: scripts/index_to_sqlite3.sh (hunk 2 above)
Comment (wbazant) [reply scripts/index_to_sqlite3.sh:1]: Will add the comment! The .fai is pretty much a table, yes.

File: src/main/java/org/veupathdb/service/sr/reference/FakeFastaSequenceIndex.java (hunk 4 above)
Comment (wbazant) [reply src/main/java/org/veupathdb/service/sr/reference/FakeFastaSequenceIndex.java]: Will add a comment, and rename to FastaSequenceIndexStub

File: src/main/java/org/veupathdb/service/sr/response/StreamSequences.java (hunk 5 above)
Comment (wbazant) [reply src/main/java/org/veupathdb/service/sr/response/StreamSequences.java:27]: It could be excessively long! It won't reach this stage, but I think the features will still all go into memory with the current implementation. Actually it's something to check - make a very long list of features, and make sure it's rejected without blowing up the container - I'll make an issue for it.

Comment (wbazant) [reply src/main/java/org/veupathdb/service/sr/response/StreamSequences.java:27]: https://github.com/VEuPathDB/service-sequence-retrieval/issues/2#issuecomment-1315594898

File: src/main/java/org/veupathdb/service/sr/response/StreamSequences.java (hunk 7 above)
Comment (wbazant) [reply src/main/java/org/veupathdb/service/sr/response/StreamSequences.java:32]: Up to one chromosome in length - so a few megabytes tops - and it would be hard to improve this, because "reverse & complement" currently transposes the whole feature instead of reading backwards.

File: src/main/java/org/veupathdb/service/sr/service/SequenceRetrievalAsyncService.java (hunk 9 above)
Comment (wbazant) [reply src/main/java/org/veupathdb/service/sr/service/SequenceRetrievalAsyncService.java:95]: This one can totally blow up, it should actually not be an endpoint - only the URL endpoint will let us stream the sequences in the async version. Let's remove what we need to remove once it's clearer!

File: src/main/java/org/veupathdb/service/sr/reference/ReferenceDAO.java (hunk 3 above)
Comment (wbazant) [reply src/main/java/org/veupathdb/service/sr/reference/ReferenceDAO.java]: Right, the ResultSet is not being closed!

**Comment by wbazant** [comment]
Pushed changes into main!
